server/pairing/client.go

```diff
@@ -210,6 +192,12 @@ func (c *SenderClient) receiveInstallationData() error {
 		return err
 	}
 
+	err = c.challengeTaker.DoChallenge(req)
```

Comment:
maybe we can hide the detail of doing challenge in `BaseClient` so that we only need focus on which endpoint we need interact and what real `business data` we need exchange? 🤔

Reply:
I like this idea I've been trying to figure out how we could do this. The difficulty I am having is how do we pass the new request to the base client before the request is instantiated by the receiving / sending client?

What sort of approach did you have in mind?

Reply:
My initial idea is to add a method called DoChallengeRequest to the SenderClient to wrap the Client.Do method. The rough code is as follows:
```
func (c *SenderClient) DoChallengeRequest(req *Request) (*Response, error) {
    err = c.challengeTaker.DoChallenge(req)
    if err != nil {
        signal.SendLocalPairingEvent(Event{Type: EventTransferError, Error: err.Error(), Action: ActionPairingInstallation})
        return nil, err
    }
    c.Do(req) ...
}
```

is this a good start?

Reply:
Ah, thank you @qfrank for your suggestion. I think that this change is getting into the area of generalising `Client` logic rather than updating the Challenge management.

However I did initially work with this idea during the development of #3248 . It is cool to know that we both broadly agree on this idea.

I think that we could generalise so that the `BaseClient` has 3 main functions `sendData`, `receiveData` and `exchangeData`.

I'll define `exchangeData` later, I need to think about how that could work.

**`sendData`**

```go
func (c *BaseClient) sendData(m PayloadMounter, action Action, path, description string) error {
	err := m.Mount()
	if err != nil {
		return err
	}

	c.baseAddress.Path = path
	req, err := http.NewRequest(http.MethodPost, c.baseAddress.String(), bytes.NewBuffer(m.ToSend()))
	if err != nil {
		return err
	}
	req.Header.Set("Content-Type", "application/octet-stream")

	err = c.challengeTaker.DoChallenge(req)
	if err != nil {
		signal.SendLocalPairingEvent(Event{Type: EventTransferError, Error: err.Error(), Action: action})
		return err
	}

	resp, err := c.Do(req)
	if err != nil {
		signal.SendLocalPairingEvent(Event{Type: EventTransferError, Error: err.Error(), Action: action})
		return err
	}

	if resp.StatusCode != http.StatusOK {
		err = fmt.Errorf("[client] status not okay when sending %s, status: %s", description, resp.Status)
		signal.SendLocalPairingEvent(Event{Type: EventTransferError, Error: err.Error(), Action: action})
		return err
	}

	signal.SendLocalPairingEvent(Event{Type: EventTransferSuccess, Action: action})
	return nil
}
```

**`receiveData`**

```go
func (c *BaseClient) receiveData(r PayloadReceiver, path string, action Action, description string) error {
	c.baseAddress.Path = path
	req, err := http.NewRequest(http.MethodGet, c.baseAddress.String(), nil)
	if err != nil {
		return err
	}

	err = c.challengeTaker.DoChallenge(req)
	if err != nil {
		signal.SendLocalPairingEvent(Event{Type: EventTransferError, Error: err.Error(), Action: action})
		return err
	}

	resp, err := c.Do(req)
	if err != nil {
		signal.SendLocalPairingEvent(Event{Type: EventTransferError, Error: err.Error(), Action: action})
		return err
	}

	if resp.StatusCode != http.StatusOK {
		err = fmt.Errorf("[client] status not ok when receiving %s, received '%s'", description, resp.Status)
		signal.SendLocalPairingEvent(Event{Type: EventTransferError, Error: err.Error(), Action: action})
		return err
	}

	payload, err := io.ReadAll(resp.Body)
	if err != nil {
		signal.SendLocalPairingEvent(Event{Type: EventTransferError, Error: err.Error(), Action: action})
		return err
	}
	signal.SendLocalPairingEvent(Event{Type: EventTransferSuccess, Action: action})

	err = r.Receive(payload)
	if err != nil {
		signal.SendLocalPairingEvent(Event{Type: EventProcessError, Error: err.Error(), Action: action})
		return err
	}
	signal.SendLocalPairingEvent(Event{Type: EventProcessSuccess, Action: action})
	return nil
}
```

Change all the `Sender` and `Receiver` funcs like below:

```go
func (c *SenderClient) sendAccountData() error {
	return c.sendData(c.accountMounter, ActionPairingAccount, pairingReceiveAccount, "account data")
}
```

What I'll do is add an issue from this comment and implement the full solutions in a dedicated PR.